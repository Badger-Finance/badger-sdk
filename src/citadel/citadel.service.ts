import {
  CitadelMinter,
  CitadelMinter__factory,
  StakedCitadelLocker,
  StakedCitadelLocker__factory,
  SupplySchedule,
  SupplySchedule__factory,
} from '..';
import { BigNumberish, ethers } from 'ethers';
import { Network } from '../config';
import { BadgerSDK } from '../sdk';
import { Service } from '../service';
import { chunkQueryFilter } from '../utils/chunk-query-filter';
import {
  CitadelDistributionToStakingEvent,
  CitadelDistributionToStakingEventFilter,
} from '../contracts/CitadelMinter';
import { evaluateDistributionEvents, parseTypedEvents } from './citadel.utils';
import { ListDistributionOptions } from './interfaces/list-distribution-options.interface';
import { ListRewardsOptions } from './interfaces/list-rewards-options.interface';
import { RewardFilter } from './enums/reward-filter.enum';
import { ListRewardsEvent } from './interfaces/list-rewards-event.interface';
import {
  RewardAddedEvent,
  RewardAddedEventFilter,
  RewardPaidEvent,
  RewardPaidEventFilter,
} from '../contracts/StakedCitadelLocker';
import { CitadelMintDistribution } from './interfaces/citadel-mint-distribution.interface';
import { getBlockDeployedAt } from '../utils/deployed-at.util';
import { RewardEventTypeEnum } from './enums/reward-event-type.enum';

export const citadelMinterAddress =
  '0x594691aEa75080dd9B3e91e648Db6045d4fF6E22';
export const stakedCitadelLockerAddress =
  '0xB1c38253aD6Ab3e2A2D53A094692fcf1321b12d4';

export class CitadelService extends Service {
  private minter?: CitadelMinter;
  private locker?: StakedCitadelLocker;

  constructor(sdk: BadgerSDK) {
    super(sdk);
    this.#init();
  }

  get citadelMinter(): CitadelMinter {
    if (!this.minter) {
      throw new Error(`Minter is not defined for ${this.config.network}`);
    }
    return this.minter;
  }

  get citadelLocker(): StakedCitadelLocker {
    if (!this.locker) {
      throw new Error(`Locker is not defined for ${this.config.network}`);
    }
    return this.locker;
  }

  async getSupplySchedule(): Promise<SupplySchedule> {
    return SupplySchedule__factory.connect(
      await this.citadelMinter.supplySchedule(),
      this.provider,
    );
  }

  getLastMintTimestamp() {
    return this.citadelMinter.lastMintTimestamp();
  }

  async listDistributions(options: ListDistributionOptions = {}) {
    const deployedAt = getBlockDeployedAt(
      citadelMinterAddress,
      this.config.network,
    );

    const { startBlock, endBlock } = options;

    if (!startBlock) options.startBlock = deployedAt;
    if (!endBlock) options.endBlock = await this.sdk.provider.getBlockNumber();

    const distFilter =
      this.citadelMinter.filters.CitadelDistributionToStaking();

    const distributionsToStaking = await chunkQueryFilter<
      CitadelMinter,
      CitadelDistributionToStakingEventFilter,
      CitadelDistributionToStakingEvent
    >(
      this.citadelMinter,
      distFilter,
      <number>options.startBlock,
      <number>options.endBlock,
    );

    const distributions = await parseTypedEvents(
      distributionsToStaking,
      (e, b) => ({
        block: b.number,
        startTime: e.args[0].toNumber(),
        endTime: e.args[1].toNumber(),
        citadelAmount: e.args[2],
      }),
    );

    return evaluateDistributionEvents(distributions, options);
  }

  isDistributor(token: string, distributor: string) {
    if (!this.citadelLocker) {
      throw new Error(`Locker not defined for ${this.config.network}`);
    }

    const tokenAddr = ethers.utils.getAddress(token);
    const distributorAddr = ethers.utils.getAddress(distributor);

    return this.citadelLocker.rewardDistributors(tokenAddr, distributorAddr);
  }

  getRewardStats(account: string) {
    const accAddr = ethers.utils.getAddress(account);

    return this.citadelLocker.rewardData(accAddr);
  }

  getRewardTokens() {
    return this.citadelLocker.getRewardTokens();
  }

  async listRewards(options: ListRewardsOptions = {}) {
    const deployedAt = getBlockDeployedAt(
      stakedCitadelLockerAddress,
      this.config.network,
    );

    const {
      startBlock,
      endBlock,
      user,
      token,
      filter = RewardFilter.ADDED,
    } = options;

    if (!startBlock) options.startBlock = deployedAt;
    if (!endBlock) options.endBlock = await this.sdk.provider.getBlockNumber();

    let rewardEvents: ListRewardsEvent[] = [];

    switch (filter) {
      case RewardFilter.ADDED:
        const addedFilter = this.citadelLocker.filters.RewardAdded(null, token);

        const addedRewardEvents = await chunkQueryFilter<
          StakedCitadelLocker,
          RewardAddedEventFilter,
          RewardAddedEvent
        >(
          this.citadelLocker,
          addedFilter,
          <number>options.startBlock,
          <number>options.endBlock,
        );

        rewardEvents = await parseTypedEvents(addedRewardEvents, (e, b) => ({
          block: b.number,
          account: e.args[0],
          token: e.args[1],
          reward: e.args[2],
          dataTypeHash: e.args[3],
          timestamp: e.args[4].toNumber(),
          type: RewardEventTypeEnum.ADDED,
        }));
        break;
      case RewardFilter.PAID:
        const paidFilter = this.citadelLocker.filters.RewardPaid(user, token);

        const paidRewardEvents = await chunkQueryFilter<
          StakedCitadelLocker,
          RewardPaidEventFilter,
          RewardPaidEvent
        >(
          this.citadelLocker,
          paidFilter,
          <number>options.startBlock,
          <number>options.endBlock,
        );

        rewardEvents = await parseTypedEvents(paidRewardEvents, (e, b) => ({
          block: b.number,
          user: e.args[0],
          token: e.args[1],
          reward: e.args[2],
          type: RewardEventTypeEnum.PAID,
        }));
        break;
      default:
        throw new Error(`Unknown reward filter ${filter}`);
    }

    return rewardEvents;
  }

  async getClaimableRewards(account: string) {
    const accAddr = ethers.utils.getAddress(account);
    return this.citadelLocker.claimableRewards(accAddr);
  }

  async rewardWeightOf(account: string) {
    const accAddr = ethers.utils.getAddress(account);
    return this.citadelLocker.rewardWeightOf(accAddr);
  }

  lockedBalanceOf(address: string) {
    const userAddr = ethers.utils.getAddress(address);
    return this.citadelLocker.lockedBalanceOf(userAddr);
  }

  balanceOf(address: string) {
    const userAddr = ethers.utils.getAddress(address);
    return this.citadelLocker.balanceOf(userAddr);
  }

  balanceAtEpochOf(epoch: BigNumberish, address: string) {
    const userAddr = ethers.utils.getAddress(address);
    return this.citadelLocker.balanceAtEpochOf(epoch, userAddr);
  }

  getEcochs(index: BigNumberish) {
    return this.citadelLocker.epochs(index);
  }

  getLockedSupply() {
    return this.citadelLocker.lockedSupply();
  }

  getBoostedSupply() {
    return this.citadelLocker.boostedSupply();
  }

  async getCitadelMintDistribution(): Promise<CitadelMintDistribution> {
    const [fundingBps, stakingBps, lockingBps] = await Promise.all([
      this.citadelMinter.fundingBps(),
      this.citadelMinter.stakingBps(),
      this.citadelMinter.lockingBps(),
    ]);

    return {
      fundingBps: fundingBps.toNumber(),
      stakingBps: stakingBps.toNumber(),
      lockingBps: lockingBps.toNumber(),
    };
  }

  #init() {
    if (this.config.network !== Network.Ethereum) {
      return;
    }
    this.minter = CitadelMinter__factory.connect(
      citadelMinterAddress,
      this.provider,
    );
    this.locker = StakedCitadelLocker__factory.connect(
      stakedCitadelLockerAddress,
      this.provider,
    );
  }
}
