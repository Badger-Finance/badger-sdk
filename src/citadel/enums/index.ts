export * from './reward-data-type.enum';
export * from './reward-event-type.enum';
export * from './reward-filter.enum';
