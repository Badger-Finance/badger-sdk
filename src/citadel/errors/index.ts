export { CitadelInitError } from './citadel.init.error';
export * as errCodes from './citadel.init.error';
export { CitadelValidationError } from './citadel.validation.error';
