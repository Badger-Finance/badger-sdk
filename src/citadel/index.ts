export { CitadelService } from './citadel.service';
