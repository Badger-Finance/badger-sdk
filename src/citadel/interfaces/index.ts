export { CitadelMintDistribution } from './citadel-mint-distribution.interface';
export { ListDistributionOptions } from './list-distribution-options.interface';
export { ListRewardsEvent } from './list-rewards-event.interface';
export { ListRewardsOptions } from './list-rewards-options.interface';
export { MinterDistributionData } from './minter-distribution-data.interface';
export { MinterDistributionEvent } from './minter-distribution-event.interface';
