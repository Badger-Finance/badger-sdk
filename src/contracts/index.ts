/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type { BadgerTree } from "./BadgerTree";
export type { Digg } from "./Digg";
export type { Erc20 } from "./Erc20";
export type { Ibbtc } from "./Ibbtc";
export type { Registry } from "./Registry";
export type { RewardsLogger } from "./RewardsLogger";
export type { Sett } from "./Sett";
export type { StrategyV15 } from "./StrategyV15";
export type { VaultV15 } from "./VaultV15";

export { BadgerTree__factory } from "./factories/BadgerTree__factory";
export { Digg__factory } from "./factories/Digg__factory";
export { Erc20__factory } from "./factories/Erc20__factory";
export { Ibbtc__factory } from "./factories/Ibbtc__factory";
export { Registry__factory } from "./factories/Registry__factory";
export { RewardsLogger__factory } from "./factories/RewardsLogger__factory";
export { Sett__factory } from "./factories/Sett__factory";
export { StrategyV15__factory } from "./factories/StrategyV15__factory";
export { VaultV15__factory } from "./factories/VaultV15__factory";
