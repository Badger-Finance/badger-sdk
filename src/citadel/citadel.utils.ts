import { CitadelDistributionToStakingEvent } from '../contracts/CitadelMinter';
import { MinterDistributionEvent } from './interfaces/minter-distribution-event.interface';
import { TimeRangeOptions } from '../common';
import { MinterDistributionData } from './interfaces/minter-distribution-data.interface';
import { timestampInRange } from '../vaults/vaults.utils';
import { keyBy } from '../utils';
import { ListRewardsEvent } from './interfaces/list-rewards-event.interface';
import {
  RewardAddedEvent,
  RewardPaidEvent,
} from '../contracts/StakedCitadelLocker';

export function parseDistributionsEvents(
  events: CitadelDistributionToStakingEvent[],
): Promise<MinterDistributionEvent[]> {
  return Promise.all(
    events.map(async (e) => {
      const block = await e.getBlock();

      return {
        block: block.number,
        startTime: e.args[0].toNumber(),
        endTime: e.args[1].toNumber(),
        citadelAmount: e.args[2],
      };
    }),
  );
}

export function evaluateDistributionEvents<T extends TimeRangeOptions>(
  events: MinterDistributionEvent[],
  options: T,
): MinterDistributionData {
  const filteredEvents = events.filter((h) =>
    timestampInRange(options, h.startTime),
  );

  const distTimeMap = keyBy(filteredEvents, (event) => event.startTime);

  const timestamps = Array.from(new Set([...distTimeMap.keys()])).sort();

  const distTimeOrderedMap: MinterDistributionData = {};

  for (const time of timestamps) {
    distTimeOrderedMap[`${time}`] = distTimeMap.get(time) ?? [];
  }

  return distTimeOrderedMap;
}

export function parseAddedRewardEvents(
  events: RewardAddedEvent[],
): Promise<ListRewardsEvent[]> {
  return Promise.all(
    events.map(async (e) => {
      const block = await e.getBlock();

      return {
        block: block.number,
        token: e.args[0],
        reward: e.args[1],
      };
    }),
  );
}

export function parsePaidRewardEvents(
  events: RewardPaidEvent[],
): Promise<ListRewardsEvent[]> {
  return Promise.all(
    events.map(async (e) => {
      const block = await e.getBlock();

      return {
        block: block.number,
        user: e.args[0],
        token: e.args[1],
        reward: e.args[2],
      };
    }),
  );
}
