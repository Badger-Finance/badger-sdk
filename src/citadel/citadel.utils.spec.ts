import { CitadelDistributionToStakingEvent } from '../contracts/CitadelMinter';
import { evaluateDistributionEvents, parseTypedEvents } from './citadel.utils';
import { distributionFormattedEvents } from './mocks/distribution-formatted-events.mock';
import { citadelDistributionToStakingEvents } from './mocks/distribution-to-staking-events.mock';

// TypedEvents require to implement all the methods in interface,
// so we ignore mocks for testing purposes

describe('parseTypedEvents', () => {
  it('should give formatted events', async () => {
    expect(
      await parseTypedEvents(
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        citadelDistributionToStakingEvents,
        (e, b) => ({
          block: b.number,
          startTime: e.args[0].toNumber(),
          endTime: e.args[1].toNumber(),
          citadelAmount: e.args[2],
        }),
      ),
    ).toMatchSnapshot();
  });

  it('should return empty list', async () => {
    expect(
      (
        await parseTypedEvents(
          <CitadelDistributionToStakingEvent[]>[],
          (e, b) => ({
            block: b.number,
            startTime: e.args[0].toNumber(),
            endTime: e.args[1].toNumber(),
            citadelAmount: e.args[2],
          }),
        )
      ).length,
    ).toBe(0);
  });
});

describe('evaluateDistributionEvents', () => {
  it('should map events by timestamp', () => {
    expect(
      evaluateDistributionEvents(distributionFormattedEvents, {}),
    ).toMatchSnapshot();
  });

  it('should map events by timestamp, and filter them', () => {
    expect(
      evaluateDistributionEvents(distributionFormattedEvents, {
        timestamp_gte: 1650712126545,
      }),
    ).toMatchSnapshot();
  });

  it('should return empty list', () => {
    expect(Object.keys(evaluateDistributionEvents([], {})).length).toBe(0);
  });
});
