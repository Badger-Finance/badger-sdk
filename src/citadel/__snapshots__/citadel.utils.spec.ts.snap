// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evaluateDistributionEvents should map events by timestamp 1`] = `
Object {
  "1650711953853": Array [
    Object {
      "block": 10,
      "citadelAmount": Object {
        "hex": "0x64",
        "type": "BigNumber",
      },
      "endTime": 1650712062667,
      "startTime": 1650711953853,
    },
    Object {
      "block": 10,
      "citadelAmount": Object {
        "hex": "0x32",
        "type": "BigNumber",
      },
      "endTime": 1650712062667,
      "startTime": 1650711953853,
    },
    Object {
      "block": 10,
      "citadelAmount": Object {
        "hex": "0x0f41fe",
        "type": "BigNumber",
      },
      "endTime": 1650712062667,
      "startTime": 1650711953853,
    },
  ],
  "1650712126545": Array [
    Object {
      "block": 11,
      "citadelAmount": Object {
        "hex": "0x0f41fe",
        "type": "BigNumber",
      },
      "endTime": 1650712135672,
      "startTime": 1650712126545,
    },
    Object {
      "block": 11,
      "citadelAmount": Object {
        "hex": "0x020c1181",
        "type": "BigNumber",
      },
      "endTime": 1650712135672,
      "startTime": 1650712126545,
    },
  ],
  "1650712161976": Array [
    Object {
      "block": 12,
      "citadelAmount": Object {
        "hex": "0x020c1181",
        "type": "BigNumber",
      },
      "endTime": 1650712167379,
      "startTime": 1650712161976,
    },
  ],
}
`;

exports[`evaluateDistributionEvents should map events by timestamp, and filter them 1`] = `
Object {
  "1650712126545": Array [
    Object {
      "block": 11,
      "citadelAmount": Object {
        "hex": "0x0f41fe",
        "type": "BigNumber",
      },
      "endTime": 1650712135672,
      "startTime": 1650712126545,
    },
    Object {
      "block": 11,
      "citadelAmount": Object {
        "hex": "0x020c1181",
        "type": "BigNumber",
      },
      "endTime": 1650712135672,
      "startTime": 1650712126545,
    },
  ],
  "1650712161976": Array [
    Object {
      "block": 12,
      "citadelAmount": Object {
        "hex": "0x020c1181",
        "type": "BigNumber",
      },
      "endTime": 1650712167379,
      "startTime": 1650712161976,
    },
  ],
}
`;

exports[`parseTypedEvents should give formatted events 1`] = `
Array [
  Object {
    "block": 50,
    "citadelAmount": Object {
      "hex": "0x1388",
      "type": "BigNumber",
    },
    "endTime": 1650726976504,
    "startTime": 1650726066923,
  },
  Object {
    "block": 50,
    "citadelAmount": Object {
      "hex": "0x0159",
      "type": "BigNumber",
    },
    "endTime": 1650726976509,
    "startTime": 1650726066923,
  },
  Object {
    "block": 52,
    "citadelAmount": Object {
      "hex": "0x087a22",
      "type": "BigNumber",
    },
    "endTime": 1650727116602,
    "startTime": 1650727090814,
  },
  Object {
    "block": 52,
    "citadelAmount": Object {
      "hex": "0x0c",
      "type": "BigNumber",
    },
    "endTime": 1650727116699,
    "startTime": 1650727090814,
  },
  Object {
    "block": 55,
    "citadelAmount": Object {
      "hex": "0x0c",
      "type": "BigNumber",
    },
    "endTime": 1650727116699,
    "startTime": 1650727090999,
  },
]
`;
